src/urllib3/response.py

```diff
@@ -701,14 +702,46 @@ def read(
         flush_decoder = False
         fp_closed = getattr(self._fp, "closed", False)
 
+        # Reading more than 2 GiB (`int` max value in C) via SSL on some
+        # CPython 3.8, 3.9, and 3.10 versions leads to an overflow error
+        # that has to be prevented if `amt` or `self.length_remaining`
+        # indicate that it may happen.
+        c_int_max = (2**31) - 1
+        if (
+            (amt and amt > c_int_max)
+            or (self.length_remaining and self.length_remaining > c_int_max)
+        ) and (
+            (3, 8, 0, "a", 4) <= sys.version_info < (3, 9, 7)
+            or (3, 10, 0, "b", 1) <= sys.version_info < (3, 10, 0, "rc", 1)
+        ):
+
+            def read(amt: Optional[int] = None) -> bytes:
+                if self._fp is None:
+                    return b""
+                buffer = io.BytesIO()
+                while amt is None or amt != 0:
+                    if amt is not None:
+                        chunk_amt = min(amt, c_int_max)
+                        amt -= chunk_amt
+                    else:
+                        chunk_amt = c_int_max
+                    data = self._fp.read(chunk_amt)
+                    if not data:
+                        break
+                    buffer.write(data)
+                return buffer.getvalue()
```

Comment:
OK so we're talking about gigabytes of memory here. For a 3GB read, what is the peak memory usage? Is there a way to ensure we never use more than 3GB of memory?

Reply:
If the content is 3GB and the user calls `resp.read()` we have to load the whole 3GB into memory, there's no way around that unfortunately?

Reply:
Sure but it's easy to accidentally use say 6GB instead.

Reply:
Ah now I understand what you meant, wasn't thinking about overhead!

Reply:
Is this a good reason to use smaller chunks?

Reply:
Good idea! The smaller the chunk, the smaller the overhead should be.

I would not make it too small though. Maybe 512 MiB?

Python used to read in 1 MiB-chunks. Obviously, there were complaints https://github.com/python/cpython/issues/80231.